### ISSUE TITLE
Model Instances Not Deploying in v3

### ISSUE THREAD

**Issue by david-saeger** [awslabs/LISA#101]
After tearing down all v2 infrastructure and re deploying. All Stacks report CREATE_COMPLETE status. After deploying none of the models I have in configuration have compute instances attached and there are no registeredModels.

My config looks like this 

```yaml
env: dev
dev:
  appName: lisa
  deploymentName: owl
  accountNumber: ######
  region: #######
  deploymentStage: dev
  removalPolicy: destroy
  runCdkNag: false
  systemBanner:
    text: 'Lisa'
    backgroundColor: blue
    fontColor: white
  s3BucketModels: owl-models
  mountS3DebUrl: https://s3.amazonaws.com/mountpoint-s3-release/latest/x86_64/mount-s3.deb
  accountNumbersEcr:
    - #######
  deployRag: true
  lambdaConfig:
    pythonRuntime: PYTHON_3_10
    logLevel: DEBUG
    vpcAutoscalingConfig:
      provisionedConcurrentExecutions: 5
      minCapacity: 1
      maxCapacity: 50
      targetValue: 0.80
      cooldown: 30
  authConfig:
    authority: https://cognito-idp.us-gov-west-1.amazonaws.com/us-gov-west-1_iG9gEGqj1
    clientId: 3ag8sc91egsi4f12kju3lfdf1p
    adminGroup: admin
    jwtGroupsProperty: cognito:groups
  logLevel: DEBUG
  restApiConfig:
    apiVersion: v2
    instanceType: m5.large
    containerConfig:
      image:
        baseImage: python:3.9
        path: lib/serve/rest-api
        type: asset
      healthCheckConfig:
        command: ['CMD-SHELL', 'exit 0']
        interval: 10
        startPeriod: 30
        timeout: 5
        retries: 3
    autoScalingConfig:
      minCapacity: 1
      maxCapacity: 1
      cooldown: 60
      defaultInstanceWarmup: 60
      metricConfig:
        AlbMetricName: RequestCountPerTarget
        targetValue: 1000
        duration: 60
        estimatedInstanceWarmup: 30
    loadBalancerConfig:
      sslCertIamArn: arn:aws-us-gov:iam::189215246007:server-certificate/owl-dev
      healthCheckConfig:
        path: /health
        interval: 60
        timeout: 30
        healthyThresholdCount: 2
        unhealthyThresholdCount: 10
      domainName:
  ragRepositories:
    - repositoryId: pgvector-rag
      type: pgvector
      rdsConfig:
        username: postgres
 
  ragFileProcessingConfig:
    chunkSize: 512
    chunkOverlap: 51
  ecsModels:
    - modelName: mistralai/Mistral-7B-Instruct-v0.2
      modelId: mistral7b
      deploy: true
      streaming: true
      modelType: textgen
      instanceType: g4dn.2xlarge
      inferenceContainer: tgi
      containerConfig:
        image:
          baseImage: ghcr.io/huggingface/text-generation-inference:2.0.4
          path: lib/serve/ecs-model/textgen/tgi
          type: asset
        sharedMemorySize: 2048
        healthCheckConfig:
          command: ['CMD-SHELL', 'exit 0']
          interval: 10
          startPeriod: 30
          timeout: 5
          retries: 3
        environment:
          MAX_CONCURRENT_REQUESTS: 128
          MAX_INPUT_LENGTH: 1024
          MAX_TOTAL_TOKENS: 2048
          RUST_BACKTRACE: 1
      autoScalingConfig:
        minCapacity: 1
        maxCapacity: 1
        defaultInstanceWarmup: 180
        cooldown: 420
        metricConfig:
          AlbMetricName: RequestCountPerTarget
          targetValue: 30
          duration: 60
          estimatedInstanceWarmup: 330
      loadBalancerConfig:
        healthCheckConfig:
          path: /health
          interval: 60
          timeout: 30
          healthyThresholdCount: 2
          unhealthyThresholdCount: 10


    - modelName: intfloat/e5-large-v2
      modelId: e5v2
      deploy: true
      modelType: embedding
      instanceType: g4dn.xlarge
      inferenceContainer: tei
      containerConfig:
        image:
          baseImage: ghcr.io/huggingface/text-embeddings-inference:turing-latest
          path: lib/serve/ecs-model/embedding/tei
          type: asset
        sharedMemorySize: 2048
        healthCheckConfig:
          command: ['CMD-SHELL', 'exit 0']
          interval: 10
          startPeriod: 30
          timeout: 5
          retries: 3
        environment:
          MAX_CONCURRENT_REQUESTS: 512
          MAX_CLIENT_BATCH_SIZE: 1024
          MAX_BATCH_TOKENS: 16384
          RUST_BACKTRACE: 1
      autoScalingConfig:
        minCapacity: 1
        maxCapacity: 1
        cooldown: 420
        defaultInstanceWarmup: 180
        metricConfig:
          AlbMetricName: RequestCountPerTarget
          targetValue: 60
          duration: 60
          estimatedInstanceWarmup: 330
      loadBalancerConfig:
        healthCheckConfig:
          path: /health
          interval: 30
          timeout: 10
          healthyThresholdCount: 2
          unhealthyThresholdCount: 10
 
  litellmConfig:
    litellm_settings:
      telemetry: false # Don't try to send telemetry to LiteLLM servers.
    general_settings:
      master_key: xxxxxxxxxxxx1 # A key is required for model management purposes
```

**Comment by petermuller** [comment]
Hi David, unfortunately, that is expected- we called that out here as part of the v2 -> v3 migration docs [here](https://github.com/awslabs/LISA?tab=readme-ov-file#breaking-changes-in-v2-to-v3-migration)

The good news though is with the configuration of an "admin" group in your IDP, you will then have access to the Model Management tab for adding the models with these same params (except the `path` component of `containerConfig.image.path`, which after scanning the readme file, i see was removed 😢 ) That behavior here, outlining that the "path" is now relative to the ecs-model directory instead of the repository root: https://github.com/awslabs/LISA/commit/8a08f9bb45fa7dfef4e49a842988c8fda03be1a7#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R56-R60

and I'll get a new PR out to add that information back because that's an important one 😅

**Comment by petermuller** [comment]
Added as part of the tagged commit and now referenced here https://github.com/awslabs/LISA?tab=readme-ov-file#creating-a-model-admin-api

**Comment by david-saeger** [comment]
Ahh darn. Should have read the docs closer! Apologies!

**Comment by david-saeger** [comment]
can confirm I was able to add a model through the API. Strangely the UI is not showcasing a model management tab despite the fact that I have updated added a user group "admin" and set it in authconfig in the LISA configuration

```
  authConfig:
    authority: xxxxx
    clientId: xxxxx
    adminGroup: admin
    jwtGroupsProperty: cognito:groups
```

This is not a blocker for me but just noting in case it is a broader issue or perhaps I am missing some info on how to navigate to the administrative panel.

**Comment by david-saeger** [comment]
I may have spoken too soon. I created a model through the API which reports back from get [api gateway]/models as 

```
{
    "models": [
        {
            "autoScalingConfig": {
                "minCapacity": 1,
                "maxCapacity": 1,
                "cooldown": 420,
                "defaultInstanceWarmup": 180,
                "metricConfig": {
                    "albMetricName": "RequestCountPerTarget",
                    "targetValue": 30,
                    "duration": 60,
                    "estimatedInstanceWarmup": 330
                }
            },
            "containerConfig": {
                "baseImage": {
                    "baseImage": "vllm/vllm-openai:v0.5.0",
                    "path": "vllm",
                    "type": "asset"
                },
                "sharedMemorySize": 2048,
                "healthCheckConfig": {
                    "command": [
                        "CMD-SHELL",
                        "exit 0"
                    ],
                    "interval": 10,
                    "startPeriod": 30,
                    "timeout": 5,
                    "retries": 3
                },
                "environment": {
                    "MAX_TOTAL_TOKENS": "2048",
                    "MAX_CONCURRENT_REQUESTS": "128",
                    "MAX_INPUT_LENGTH": "1024"
                }
            },
            "loadBalancerConfig": {
                "healthCheckConfig": {
                    "path": "/health",
                    "interval": 60,
                    "timeout": 30,
                    "healthyThresholdCount": 2,
                    "unhealthyThresholdCount": 10
                }
            },
            "instanceType": "g4dn.2xlarge",
            "modelId": "mistral-vllm",
            "modelName": "mistralai/Mistral-7B-Instruct-v0.2",
            "modelType": "textgen",
            "modelUrl": "",
            "status": "InService",
            "streaming": true
        }
    ]
}
```

But don't see any instances running in it other than the RestAPI.

**Comment by david-saeger** [comment]
I see a bunch of stuff in 3.0.1 that might have caught an error in create API which might have given me more info. will delete for now and wait on that.

**Comment by david-saeger** [comment]
When creating a model through the API does the model id need to match the object in s3 which stores safetensors? Not quite sure I grok how the new model management endpoint is intended to interact with the location of the model itself within s3. Or is it the case that a new model creation request will pull the appropriate artifacts from huggingface

**Comment by david-saeger** [comment]
Tried pulling in some recent commits redeployed and tried again. Now the model is just stuck in `Creating`

**Comment by petermuller** [comment]
Actions for LISA team:
* better documentation / explanation of fields

---

So the "modelId" is just the human friendly name that will be shown as part of the OpenAI spec call (the standard /models route on the LISA Serve endpoint). The "modelName" will be the S3 key path to where your safetensors are located. So for instance, if you have the model weights stored at `s3://your-bucket/mistralai/mistral-7b`, then the "modelName" should be `mistralai/mistral-7b`. If your path is simply `s3://your-bucket/mistral-7b`, then the modelName should just be `mistral-7b`. 

The bucket where the artifacts will live is still defined in the root config.yaml file, under the `s3BucketModels` option here: https://github.com/awslabs/LISA/blob/develop/lib/schema.ts#L876

based on the output you had from <apigw>/models, it seems like almost everything worked except for setting the model url, which without that, LiteLLM doesn't know how to reference the model, so it doesn't seem like it ever got added to the db.

if you're still stuck in "Creating", you can check to see any failed state machine executions in stepfunctions, and I could possibly help to get it into a delete-able state. core things we'd want to check:
1. are there any extraneous resources that we need to clean up?
2. if the delete api also doesn't work, then can we edit the model's dynamodb entry to get it working right?

based on a screenshot of the state machine and where it either turned orange or red, we could probably make a good guess about what went wrong-
1. if the failure is in the ECR polling loop, then we might have exceeded our team for getting the docker image replicated into your account, so there may be a chance there's an ec2 instance to clean up if the state machine didn't already do that
2. if the failure is in the cloudformation polling loop, then we might have exceeded the time for the model to stand up (currently 30 minutes iirc). if that's the case, the cloudformation stack may still be up, so the delete model API would handle tearing it down. if delete is failing, there was a bug a while ago that the model wouldn't delete if there wasn't a LiteLLM id in it, so all you'd need to do is add an arbitrary string for the `litellm_id` field in ddb (you can just add that to an item if it's not there), and *then* the delete API should take care of the rest

**Comment by david-saeger** [comment]
This is very helpful. The Delete API worked fine so that eliminates some possibilities. The CreateModel state machine failed at AddModelToLitellm (ive attached a screenshot)
<img width="471" alt="Screenshot 2024-09-24 at 9 52 51 AM" src="https://github.com/user-attachments/assets/25a44c47-9d63-4fc5-b814-da4328db4f6e">

The state machine event error message reads 

```json
{
  "cause": {
    "errorMessage": "'model_info'",
    "errorType": "KeyError",
    "requestId": "d7086624-33c7-4087-9536-42f70c592a3d",
    "stackTrace": [
      "  File \"/var/task/models/state_machine/create_model.py\", line 256, in handle_add_model_to_litellm\n    litellm_id = litellm_response[\"model_info\"][\"id\"]\n"
    ]
  },
  "error": "KeyError",
  "resource": "invoke",
  "resourceType": "lambda"
}
```


So I take it there was no id in the litellm response https://github.com/awslabs/LISA/blob/b3d6f8fa4e2251794b5e0c20d6a4f9cd54ecc27d/lambda/models/state_machine/create_model.py#L256

The path I am entering in model name does match the path to my safetensors. I believe I had cleaned up all relevant resources before building on v3 but I may have missed something. 

Curious if there is any other information I can share that might help tracking down this bug or more likely user error.

**Comment by petermuller** [comment]
Got it, so from the state machine, it looks like it decided this was a litellm-only model and not one that warranted spinning up ECS/EC2 infrastructure. If you *did* expect EC2 infrastructure, If possible, could you post the input to the state machine? Otherwise from this point, what you can do is click on the failed state, and there should be a link to the "Log Group", which would have the Lambda logs for the AddModelToLitellm function. From there, that should have the stack trace. Since we went straight to the LiteLLM step, we also hit this condition: https://github.com/awslabs/LISA/blob/b3d6f8fa4e2251794b5e0c20d6a4f9cd54ecc27d/lambda/models/state_machine/create_model.py#L249 and if the modelName wasn't formatted in a way that LiteLLM liked, then that step could have failed and not given an error. I would have expected an exception there, but LiteLLM tends to keep doing things outside of my expectations 😅

**Comment by david-saeger** [comment]
Input to state machine: 
(looks slightly different from what I had in config in v2 in terms of base image but I just copied what I saw in the docs and updated the instance type maybe thats where I went wrong)
```
{
  "autoScalingConfig": {
    "minCapacity": 1,
    "maxCapacity": 1,
    "cooldown": 420,
    "defaultInstanceWarmup": 180,
    "metricConfig": {
      "albMetricName": "RequestCountPerTarget",
      "targetValue": 30,
      "duration": 60,
      "estimatedInstanceWarmup": 330
    }
  },
  "containerConfig": {
    "baseImage": {
      "baseImage": "vllm/vllm-openai:v0.5.0",
      "path": "vllm",
      "type": "asset"
    },
    "sharedMemorySize": 2048,
    "healthCheckConfig": {
      "command": [
        "CMD-SHELL",
        "exit 0"
      ],
      "interval": 10,
      "startPeriod": 30,
      "timeout": 5,
      "retries": 3
    },
    "environment": {
      "MAX_CONCURRENT_REQUESTS": "128",
      "MAX_INPUT_LENGTH": "1024",
      "MAX_TOTAL_TOKENS": "2048"
    }
  },
  "inferenceContainer": null,
  "instanceType": "g4dn.2xlarge",
  "loadBalancerConfig": {
    "healthCheckConfig": {
      "path": "/health",
      "interval": 60,
      "timeout": 30,
      "healthyThresholdCount": 2,
      "unhealthyThresholdCount": 10
    }
  },
  "modelId": "mistralai",
  "modelName": "mistralai/Mistral-7B-Instruct-v0.2",
  "modelType": "textgen",
  "modelUrl": null,
  "streaming": true
}
```

The link to lambda logs comes a few events before the TaskFailed event (at TaskScheduled)  with the following stacktrace 

```txt
[ERROR] KeyError: 'model_info'
Traceback (most recent call last):
  File "/var/task/models/state_machine/create_model.py", line 256, in handle_add_model_to_litellm
    litellm_id = litellm_response["model_info"]["id"]
```

**Comment by petermuller** [comment]
Ah, subtle, but one more option you need- we check for a few specific options to ensure that this will be a self-hosted model vs a litellm-only model (we should probably error out if we have ambiguous info. will make an action for that too)

```yaml
# right above your instanceType
"inferenceContainer": "vllm",
```

after this, I'd expect that it'd go through the left side of the state machine

related check: https://github.com/awslabs/LISA/blob/b3d6f8fa4e2251794b5e0c20d6a4f9cd54ecc27d/lambda/models/state_machine/create_model.py#L61-L72

**Comment by david-saeger** [comment]
ahh yes indeed! `inferenceContainer` seems to have got the state machine going down the right path.

**Comment by david-saeger** [comment]
welp made it much further so thats progress! 
<img width="802" alt="Screenshot 2024-09-24 at 3 17 49 PM" src="https://github.com/user-attachments/assets/99e94af1-e3cb-44c8-b534-e8478c29942f">

It finally errored on when polling the create stack the stack trace was not very helpful (to me) in terms of pointing to what the issue is

```
[ERROR] UnexpectedCloudFormationStateException: 
{
    "error": "Stack entered unexpected state: ROLLBACK_IN_PROGRESS",
    "event": {
        "autoScalingConfig": {
            "minCapacity": 1,
            "maxCapacity": 1,
            "cooldown": 420,
            "defaultInstanceWarmup": 180,
            "metricConfig": {
                "albMetricName": "RequestCountPerTarget",
                "targetValue": 30,
                "duration": 60,
                "estimatedInstanceWarmup": 330
            }
        },
        "containerConfig": {
            "baseImage": {
                "baseImage": "vllm/vllm-openai:v0.5.0",
                "path": "vllm",
                "type": "asset"
            },
            "sharedMemorySize": 2048,
            "healthCheckConfig": {
                "command": [
                    "CMD-SHELL",
                    "exit 0"
                ],
                "interval": 10,
                "startPeriod": 30,
                "timeout": 5,
                "retries": 3
            },
            "environment": {
                "MAX_CONCURRENT_REQUESTS": "128",
                "MAX_INPUT_LENGTH": "1024",
                "MAX_TOTAL_TOKENS": "2048"
            }
        },
        "inferenceContainer": "vllm",
        "instanceType": "g4dn.2xlarge",
        "loadBalancerConfig": {
            "healthCheckConfig": {
                "path": "/health",
                "interval": 60,
                "timeout": 30,
                "healthyThresholdCount": 2,
                "unhealthyThresholdCount": 10
            }
        },
        "modelId": "mistralai",
        "modelName": "mistralai/Mistral-7B-Instruct-v0.2",
        "modelType": "textgen",
        "modelUrl": null,
        "streaming": true,
        "create_infra": true,
        "image_info": {
            "instance_id": "i-0522eb165b2eb42bf",
            "image_tag": "ba7b5fcc-7398-47ce-8f65-1d7c4c53b5b4",
            "remaining_polls": 21
        },
        "continue_polling_docker": false,
        "stack_name": "owl-mistralai",
        "stack_arn": "arn:aws-us-gov:cloudformation:us-gov-west-1:189215246007:stack/owl-mistralai/f0f91b90-7aa3-11ef-aa91-0a11b67b1c7a",
        "remaining_polls_stack": 9,
        "continue_polling_stack": true
    }
}

Traceback (most recent call last):
  File "/var/task/models/state_machine/create_model.py", line 225, in handle_poll_create_stack
    raise UnexpectedCloudFormationStateException(
```

possible that it is just a timeout?

**Comment by petermuller** [comment]
it didn't time out on the state machine side, but it did fail in the cloudformation stack creation (hence the rolling back status). from the cloudformation console, if you look into the deleted stacks (if it's not still rolling back), you should be able to look the the cfn events to see what had caused the deployment failure. most commonly this will be an ecs circuit breaker event, in which you'd have to find the container logs in the cloudwatch console for the model you tried to spin up. if i had to guess, most likely this is getting back to a gpu compatibility issue on the g4dn's vs g5's since i think the rest of the options look correct

**Comment by david-saeger** [comment]
Thanks, I'll look into it.

**Comment by david-saeger** [comment]
tried to redeploy with an image config closer to what worked for me in v2 

```
  "modelId": "mistralai",
  "modelName": "mistralai/Mistral-7B-Instruct-v0.2",
  "modelType": "textgen",
  "inferenceContainer": "tgi",
  "instanceType": "g4dn.2xlarge",
  "containerConfig": {
    "baseImage": {
      "baseImage": "ghcr.io/huggingface/text-generation-inference:2.0.4",
      "path": "textgen/tgi",
      "type": "asset"
    },
    ...
    },
```
This time the cloudformation reports the mistralai STACK as CREATE_COMPLETE but still failing. This time at AddModelToLiteLLM
<img width="472" alt="Screenshot 2024-09-24 at 4 54 08 PM" src="https://github.com/user-attachments/assets/fde21f85-8fe5-4440-a46f-b9e55a7773f6">


The stack trace reported in the state engine is 

```
{
  "errorMessage": "'model_info'",
  "errorType": "KeyError",
  "requestId": "43ac6a48-f6be-4b21-bda2-25e730c2eefb",
  "stackTrace": [
    "  File \"/var/task/models/state_machine/create_model.py\", line 256, in handle_add_model_to_litellm\n    litellm_id = litellm_response[\"model_info\"][\"id\"]\n"
  ]
}
```

The ecs instance was indeed created but indeed litellm was not updated with this information and the model is not available from the frontend.

**Comment by petermuller** [comment]
it's odd to me that litellm isn't cooperating with you. we'll need to add more logging (and error catching) to figure out why litellm isn't accepting the calls for adding a model. in the rest api ecs container logs, can you see if there are any litellm errors thrown in there for the `/model/new` route?

**Comment by david-saeger** [comment]
Ahh ok I think I found these logs. (xxxx out uuid's)

```
2024-09-24 20:29:57 | xxxxxxxxx | ERROR    | /v2/serve/model/new | process_request | ERROR | Error occurred during processing: HTTPConnectionPool(host='localhost', port=4000): Max retries exceeded with url: /model/new (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f08f5f7b2b0>: Failed to establish a new connection: [Errno 111] Connection refused')) (took 0.15 seconds)
```
The full traceback will be hard to get in here but is the error bubbles up to application code around here 
![Screenshot 2024-09-25 at 10 47 33 AM](https://github.com/user-attachments/assets/badc0956-226b-4f6f-9f3e-bcadf82317a7)

**Comment by petermuller** [comment]
Ok, this tells me that LiteLLM never actually started up- I haven't seen that one before. In the REST API container logs (same log as before), we'd have to find when the container was initially spun up to figure out why LiteLLM failed to start. My best guesses would be something with the Postgres DB that LiteLLM needs to talk to- maybe the connection couldn't be established- or maybe with the prisma client that LiteLLM uses, which may pull external dependencies at runtime to stand up (which we've found out recently too)

**Comment by david-saeger** [comment]
welp. this isn't very helpful but I deleted the model and recreated and it built fine. Chock it up to continued LiteLLM weirdness?

**Comment by petermuller** [comment]
I see- Unclear to me why that is- the issue would have been on the REST API side (I assume?), which happens outside of the model management API related things. So unless that container was restarted again too, then I'm very unclear on why it worked, but glad that it did 😅 

Are you able to see that model / query it from the chat ui or from the LISA serve endpoint directly now?
something along the lines of the follow (`-k` flag just as a brute force to avoid the self-signed cert headaches for now)

```shell
curl -H 'Content-Type: application/json' -H "Authorization: Bearer $your_token" -k -X GET https://your-alb.amazonaws.com/v2/serve/models
```

**Comment by david-saeger** [comment]
Yes I am able to see it and query it from the ui. As far as I can tell all
is working as expected.
